src/algorithms/grasp_pr.rs

```diff
@@ -1,60 +1,38 @@
-use super::{count_forbidden_per_vertex, get_coloring_from_class_list, grasp::grasp};
+use super::{count_colors, grasp::grasp, is_coloring_valid, Solution};
 use crate::graph::Graph;
-use std::collections::HashSet;
 
-type Solution = (usize, Vec<Vec<usize>>);
+pub fn grasp_path_relinking(graph: &Graph, num_solutions_grasp: usize) -> Solution {
+    let mut solutions = grasp(graph, 10, 5, 5, num_solutions_grasp);
+    let mut best_solution = solutions.pop().unwrap().clone();
 
-pub fn grasp_path_relinking(num_iterarions_grasp: usize, graph: &Graph) -> Solution {
-    let num_vertices = graph.num_vertices();
-    let mut solutions_grasp: Vec<Solution> = Vec::new();
-    let mut best_solution: Solution = (usize::MAX, Vec::new());
-    let mut best_coloring = Vec::new();
-
-    for _ in 0..num_iterarions_grasp {
-        let new_solution = grasp(graph, 10, 5, 5);
-
-        if new_solution.0 < best_solution.0 {
-            best_solution = new_solution.clone();
-            best_coloring = get_coloring_from_class_list(num_vertices, &new_solution.1);
-        }
-
-        solutions_grasp.push(new_solution);
-    }
-
-    while let Some(mut solution) = solutions_grasp.pop() {
-        let coloring = get_coloring_from_class_list(num_vertices, &solution.1);
-        let mut difference = simmetric_difference(&best_coloring, &coloring);
+    while let Some(solution) = solutions.pop() {
+        let original_best_coloring = best_solution.1.clone();
```

Comment:
Outra coisa é que a gente está trocando a solução ótima dinamicamente, ou seja: se a gente troca a solução ótima na iteração `i`, na iteração `i+1` vamos usar essa nova solução ótima. Só que eu não sei se esse é o comportamento esperado do algoritmo, eu fiz isso isso por engano. Pra mim seria mais natural usar a solução ótima original ao longo de todo o algoritmo. O que você acha, @vinisilvag?

Reply:
De fato parece mais natural testar a solução original sempre, mas não me parece que trocar a melhor solução assim tenha tanto impacto no resultado. Tipo, se uma solução intermediária melhora, me parece ok continuar testando ela contra as soluções restantes. Por mim, pelo menos a princípio, a gente podia deixar assim e se for o caso, mais pra frente, a gente testa a implementação que fixa a solução original contra algumas poucas instâncias só pra ver se dá alguma diferença notável.